locations/jclouds/src/main/java/brooklyn/location/jclouds/JcloudsLocation.java

```diff
@@ -1305,118 +1326,148 @@ public UserCreation(LoginCredentials creds, List<Statement> statements) {
 
             // Using the pre-existing loginUser; setup the publicKey/password so can login as expected
-            if (password != null) {
-                statements.add(new ReplaceShadowPasswordEntry(Sha512Crypt.function(), loginUser, password));
-                loginCreds = LoginCredentials.builder().user(loginUser).password(password).build();
+            if (Strings.isNonBlank(passwordToSet)) {
+                statements.add(new ReplaceShadowPasswordEntry(Sha512Crypt.function(), user, passwordToSet));
+                createdUserCreds = LoginCredentials.builder().user(user).password(passwordToSet).build();
             }
-            if (publicKeyData!=null) {
-                statements.add(new AuthorizeRSAPublicKeys("~"+loginUser+"/.ssh", ImmutableList.of(publicKeyData)));
-                if (privateKeyData != null) {
-                    loginCreds = LoginCredentials.builder().user(loginUser).privateKey(privateKeyData).build();
+            if (Strings.isNonBlank(credential.getPublicKeyData())) {
+                statements.add(new AuthorizeRSAPublicKeys("~"+user+"/.ssh", ImmutableList.of(credential.getPublicKeyData())));
+                if (!credential.isUsingPassword() && Strings.isNonBlank(credential.getPrivateKeyData())) {
+                    createdUserCreds = LoginCredentials.builder().user(user).privateKey(credential.getPrivateKeyData()).build();
                 }
             }
             
-        } else if (user.equals(ROOT_USERNAME)) {
-            // Authorizes the public-key and sets password for the root user, so can login as expected
-            if (password != null) {
-                statements.add(new ReplaceShadowPasswordEntry(Sha512Crypt.function(), ROOT_USERNAME, password));
-                loginCreds = LoginCredentials.builder().user(user).password(password).build();
-            }
-            if (publicKeyData!=null) {
-                statements.add(new AuthorizeRSAPublicKeys("~"+ROOT_USERNAME+"/.ssh", ImmutableList.of(publicKeyData)));
-                if (privateKeyData != null) {
-                    loginCreds = LoginCredentials.builder().user(user).privateKey(privateKeyData).build();
+        } else {
+            String pubKey = credential.getPublicKeyData();
+            String privKey = credential.getPrivateKeyData();
+            
+            if (credential.get()==null) {
+                if (!loggedSshKeysHint && !config.containsKey(PRIVATE_KEY_FILE)) {
+                    loggedSshKeysHint = true;
+                    LOG.info("Default SSH keys not found or not usable; will create new keys for each machine. "
+                        + "Create ~/.ssh/id_rsa or "
+                        + "set "+PRIVATE_KEY_FILE.getName()+" / "+PRIVATE_KEY_PASSPHRASE.getName()+" / "+PASSWORD.getName()+" "
+                        + "as appropriate for this location if you wish to be able to log in without Brooklyn.");
                 }
+                KeyPair newKeyPair = SecureKeys.newKeyPair();
```

Comment:
This is for both use cases.  I pretty much agree with your suggestion but that could be a new feature (ie not in this PR).  All I've done here is swap the random per-machine password with a random per-machine key, because (a) it's more secure and (b) an increasing number of VM's don't support password-ssh login.

for the advanced use case, i like an explicit `create_new_key_per_machine`.  in this case i would not write it to the file system, but we make the key available through REST API or GUI.  (by the way, currently the lookup procedure is probably even worse than you expect:  we don't know what machine is being provisioned here, so if there are several machines it's going to trial-and-error to correlate.)

if this config key were not set, then in the new user case we could then create a `brooklyn_id_rsa` and `.pub` in `MGMT_BASE_DIR` (which is used subsequently as the default).  but i would only want to do this if it's just the one ssh key, and i don't want to lose the capability of new keys per machine.

have added comment to this effect.